src/hazelcore/extractions/ocaml/OCamlExtraction_Exp.re

```diff
@@ -0,0 +1,270 @@
+// (extracted expression, dynamic/inner type) like ExpandResult
+// Htyp.t is like the ground type
+
+//type extract_result = option((ocaml_exp, HTyp.t));
+type t = (ocaml_exp, HTyp.t)
+and ocaml_exp = string;
+
+/* Theorem: If extract(ctx, d) = Some(ce, ty) and ctx |- d : ty then ce :_ocaml Typ.extract(ty) */
+
+//TODO: replace == with HTyp.consistent
+let rec extract = (~ctx: Contexts.t, ~de: DHExp.t): t =>
+  switch (de) {
+  | EmptyHole(_) => failwith("Exp: Empty Hole")
```

Comment:
define an ExtractionFailed(string) exception instead of using failwith

Reply:
Oh, the reason I use `failwith` is it can directly stop the program and the string can be seen in the console. Should I write a new return type and change the program to that if we encounter a `ExtractionFailed(string)`, we just return it?